doc/configuration.md
# YAML config file template

```yaml
base:
  qredo_api_domain: play-api.qredo.network
  qredo_api_base_path: /api/v1/p
  http_scheme: https
  pin: 0
auto_approval:
  enabled: false
  retry_interval_max_sec: 300
  retry_interval_sec: 5
websocket:
  ws_scheme: wss
  reconnect_timeout_sec: 300
  reconnect_interval_sec: 5
  ping_period_sec: 5
	pong_wait_sec: 10
	write_wait_sec: 10
  read_buffer_size: 512
  write_buffer_size: 1024
http:
  addr: 0.0.0.0:8007
  cors_allow_origins:
    - '*'
  log_all_requests: false
logging:
  format: text
  level: debug
load_balancing:
  enable: false
  on_lock_error_timeout_ms: 300
  action_id_expiration_sec: 6
  redis:
    host: localhost
    port: 6379
    password: ""
    db: 0
store: 
  type: file 
  file: /volume/ccstore.db
  oci:
    compartment: ocid1.tenancy.oc1...
    vault: ocid1.vault.oc1...
    secret_encryption_key: ocid1.key.oc1...
    config_secret: automated_approver_config
  aws:
    region: aws-region-...
    config_secret: secrets_manager_secret...  
```

## Base

- **qredo_api_domain:** the domain of the api you want to use
- **qredo_api_base_path:** base path for the api urls, ex. scheme://domain/base_path
- **http_scheme:** the scheme to use for the api connection, ex. http or https
- **pin:** the pin number to use to provide a zero knowledge proof token for communication with the partner api

## Auto approval
- **enabled:** activate the automatic approval of every transaction that is received
- **retry_interval_max_sec:** the approve action maximum interval in seconds
- **retry_interval_sec:** the approve action retry interval in seconds

## Websocket
- **ws_scheme:** the scheme to use for the web socket feed connection, ex. ws or wss
- **reconnect_timeout_sec:** the reconnect timeout in seconds
- **reconnect_interval_sec:** the reconnect interval in seconds
- **ping_period_sec:** the ping period for the ping handler in seconds
- **pong_wait_sec:** the pong wait for the pong handler in seconds
- **write_wait_sec:** the write wait in seconds
- **read_buffer_size:** the websocket upgrader read buffer size in bytes
- **write_buffer_size:** the websocket upgrader write buffer size in bytes

## HTTP

- **addr:** the bind address and port the build in api endpoints
- **cors_allow_origins:** the value the the Access-Control-Allow-Origin of the responses of the build in api
- **log_all_requests:** log all incoming requests to the build in api

## Logging

- **format:** the format of the logging, ex. text or json
- **level:** the level of logging, ex. debug, info, warn, error, default is debug

## Load balancing

- **enable:** enables the load-balancing logic
- **on_lock_error_timeout_ms:** on lock timeout in milliseconds
- **action_id_expiration_sec:** expiration of action_id variable in Redis in seconds
- **redis:**
  - **host:** Redis host
  - **port:** Redis port
  - **password:** Redis password
  - **db:** Redis database name

## Store

- **type:** the type of store to use to store the private key information for the signing agent, ex. file, oci, aws
- **file:** the path to the storage file when file store is used
- **oci:** the oracle cloud configuration to store the private keys in an oracle vault
  - **compartment:** the oidc where the vault and encryption key reside
  - **vault:** the oidc of the vault where the secret will be stored
  - **secret_encryption_key:** the encryption key used for both the secret and the data inside the secret
  - **config_secret:** the name of secret that will be used to store the data
- **aws:** the amazon cloud configuration to store the private keys in amazon secrets manager
  - **region:** the AWS region where the secret is stored
  - **config_secret:** the name of the AWS Secrets Manager secret containing the encrypted data

config/config.go
package config

import (
	"os"

	"github.com/pkg/errors"
	"gopkg.in/yaml.v2"
)

// Config is the service configuration
type Config struct {
	Base          Base          `yaml:"base"`
	HTTP          HttpSettings  `yaml:"http"`
	Logging       Logging       `yaml:"logging"`
	LoadBalancing LoadBalancing `yaml:"load_balancing"`
	Store         Store         `yaml:"store"`
	AutoApprove   AutoApprove   `yaml:"auto_approval"`
	Websocket     WebSocketConf `yaml:"websocket"`
}

type Base struct {
	PIN              int    `yaml:"pin"`
	QredoAPIDomain   string `yaml:"qredo_api_domain"`
	QredoAPIBasePath string `yaml:"qredo_api_base_path"`
	HttpScheme       string `yaml:"http_scheme"`
}

type AutoApprove struct {
	Enabled          bool `yaml:"enabled"`
	RetryIntervalMax int  `yaml:"retry_interval_max_sec"`
	RetryInterval    int  `yaml:"retry_interval_sec"`
}
type WebSocketConf struct {
	WsScheme          string `yaml:"ws_scheme"`
	ReconnectTimeOut  int    `yaml:"reconnect_timeout_sec"`
	ReconnectInterval int    `yaml:"reconnect_interval_sec"`
	PingPeriod        int    `yaml:"ping_period_sec"`
	PongWait          int    `yaml:"pong_wait_sec"`
	WriteWait         int    `yaml:"write_wait_sec"`
	ReadBufferSize    int    `yaml:"read_buffer_size"`
	WriteBufferSize   int    `yaml:"write_buffer_size"`
}
type Store struct {
	Type       string    `default:"file" yaml:"type"`
	FileConfig string    `yaml:"file"`
	OciConfig  OciConfig `yaml:"oci"`
	AwsConfig  AWSConfig `yaml:"aws"`
}

type OciConfig struct {
	Compartment         string `yaml:"compartment"`
	Vault               string `yaml:"vault"`
	SecretEncryptionKey string `yaml:"secret_encryption_key"`
	ConfigSecret        string `yaml:"config_secret"`
}

// AWSConfig based signing-agent config. Used when Base StoreType is aws.
type AWSConfig struct {
	Region     string `yaml:"region"`
	SecretName string `yaml:"config_secret"`
}

type HttpSettings struct {
	Addr             string   `yaml:"addr"`
	CORSAllowOrigins []string `yaml:"cors_allow_origins"`
	LogAllRequests   bool     `yaml:"log_all_requests"`
}

type Logging struct {
	Format string `yaml:"format"`
	Level  string `yaml:"level"`
}

type LoadBalancing struct {
	Enable                bool        `yaml:"enable"`
	OnLockErrorTimeOutMs  int         `yaml:"on_lock_error_timeout_ms"`
	ActionIDExpirationSec int         `yaml:"action_id_expiration_sec"`
	RedisConfig           RedisConfig `yaml:"redis"`
}

type RedisConfig struct {
	Host     string `yaml:"host"`
	Port     int    `yaml:"port"`
	Password string `yaml:"password"`
	DB       int    `yaml:"db"`
}

// Default creates configuration with default values.
func (c *Config) Default() {
	c.HTTP = HttpSettings{
		Addr:             "127.0.0.1:8007",
		CORSAllowOrigins: []string{"*"},
	}

	c.Base.PIN = 0
	c.Base.QredoAPIDomain = "play-api.qredo.network"
	c.Base.QredoAPIBasePath = "/api/v1/p"
	c.AutoApprove = AutoApprove{
		Enabled:          false,
		RetryIntervalMax: 300,
		RetryInterval:    5,
	}
	c.Websocket = WebSocketConf{
		ReconnectTimeOut:  300,
		ReconnectInterval: 5,
		WsScheme:          "wss",
		PingPeriod:        5,
		PongWait:          10,
		WriteWait:         10,
		ReadBufferSize:    512,
		WriteBufferSize:   1024,
	}
	c.Base.HttpScheme = "https"
	c.Logging.Level = "info"
	c.Logging.Format = "json"
	c.Store.Type = "file"
	c.Store.FileConfig = "ccstore.db"
	c.LoadBalancing = LoadBalancing{
		Enable:                false,
		OnLockErrorTimeOutMs:  300,
		ActionIDExpirationSec: 6,
		RedisConfig: RedisConfig{
			Host:     "redis",
			Port:     6379,
			Password: "",
			DB:       0,
		},
	}
}

// Load reads and parses yaml config.
func (c *Config) Load(fileName string) error {
	f, err := os.ReadFile(fileName)
	if err != nil {
		return errors.Wrap(err, "read config file")
	}

	c.Default()
	if err := yaml.Unmarshal(f, c); err != nil {
		return errors.Wrap(err, "parse config file")
	}

	return nil
}

// Save saves yaml config.
func (c *Config) Save(fileName string) error {
	b, err := yaml.Marshal(c)
	if err != nil {
		return err
	}
	return os.WriteFile(fileName, b, 0600)
}
